inflate.c

```diff
@@ -914,69 +859,73 @@ int32_t Z_EXPORT PREFIX(inflate)(PREFIX3(stream) *strm, int32_t flush) {
             Tracevv((stderr, "inflate:         distance %u\n", state->offset));
             state->mode = MATCH;
 
-        case MATCH:
+        case MATCH: {
             /* copy match from window to output */
-            if (left == 0) goto inf_leave;
-            copy = out - left;
-            if (state->offset > copy) {         /* copy from window */
-                copy = state->offset - copy;
-                if (copy > state->whave) {
-                    if (state->sane) {
-                        SET_BAD("invalid distance too far back");
-                        break;
-                    }
-#ifdef INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR
-                    Trace((stderr, "inflate.c too far\n"));
-                    copy -= state->whave;
-                    copy = MIN(copy, state->length);
-                    copy = MIN(copy, left);
-                    left -= copy;
-                    state->length -= copy;
-                    do {
-                        *put++ = 0;
-                    } while (--copy);
-                    if (state->length == 0)
-                        state->mode = LEN;
-                    break;
-#endif
+            if (left == 0)
+                goto inf_leave;
+
+            unsigned char *end = state->window + (state->wsize * 2);
+            int64_t buf_left = end - put;
+            copy = state->length;
+            RESTORE();
+            if (copy > buf_left) {
+                if (strm->avail_out > 0) {
+                    /* relies on RESTORE() above with no changes to those vars */
+                    window_output_flush(strm);
+                    LOAD();
+                    buf_left = end - put;
                 }
-                if (copy > state->wnext) {
-                    copy -= state->wnext;
-                    from = state->window + (state->wsize - copy);
-                } else {
-                    from = state->window + (state->wnext - copy);
+                copy = MIN(copy, (uint32_t)buf_left);
+            }
+            if (copy == 0)
+                goto inf_leave;
+            if (state->offset > state->wnext + state->whave) {
+                if (state->sane) {
+                    SET_BAD("invalid distance too far back");
```

Comment:
This line is hit during libpng failed tests.

Reply:
Printing out the actual values would give a clue what goes wrong... It could be that either `wnext` or `whave` is not correctly updated. I'm pretty confident that `offset` is always correct, otherwise it would have failed with the original code.

Reply:
* The `wsize` is 256 because `wbits` was 8. 
* `left` variable was set to `strm->avail_out`. That was okay before, when data was copying directly to `strm->next` but not it is copied to internal buffer of size `wsize`. So now we much check to see that `left` is not greater than `strm->wsize`.
* Even with that change though, `wnext = 2, whave = 256, offset = 513`. Am I correct in thinking `offset` points to 256 + 256 = 512 which would be past the end of the pending output portion of the memory allocation?

Reply:
I think perhaps the errors from libpng are expected. 
```
inflate: allocated
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 'D'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'o'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'h'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'u'
inflate:         literal 'b'
inflate:         literal 'l'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal ' '
inflate:         literal 'd'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'a'
inflate:         literal 'i'
inflate:         literal 'n'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'r'
inflate:         literal 'e'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'i'
inflate:         literal 'v'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'm'
inflate:         literal 'o'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal ' '
inflate:         length 3
inflate:         distance 32
inflate:         literal 'e'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal '"'
inflate:         literal 'C'
inflate:         literal 'C'
inflate:         literal '0'
inflate:         literal ' '
inflate:         literal '1'
inflate:         literal '.'
inflate:         literal '0'
inflate:         literal '"'
inflate:         literal ';'
inflate:         literal ' '
inflate:         literal 'h'
inflate:         literal 't'
inflate:         literal 't'
inflate:         literal 'p'
inflate:         literal 's'
inflate:         literal ':'
inflate:         literal '/'
inflate:         literal '/'
inflate:         literal 'c'
inflate:         length 7
inflate:         distance 44
inflate:         literal 'c'
inflate:         length 6
inflate:         distance 43
inflate:         literal '.'
inflate:         literal 'o'
inflate:         literal 'r'
inflate:         literal 'g'
inflate:         literal '/'
inflate:         length 6
inflate:         distance 82
inflate:         length 6
inflate:         distance 81
inflate:         literal '/'
inflate:         literal 'z'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal 'o'
inflate:         literal '/'
inflate:         length 3
inflate:         distance 52
inflate:         literal '/'
inflate:         end of block
inflate:   check matches trailer
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 'D'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'o'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'h'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'u'
inflate:         literal 'b'
inflate:         literal 'l'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal ' '
inflate:         literal 'd'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'a'
inflate:         literal 'i'
inflate:         literal 'n'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'r'
inflate:         literal 'e'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'i'
inflate:         literal 'v'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'm'
inflate:         literal 'o'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal ' '
inflate:         length 3
inflate:         distance 32
inflate:         literal 'e'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal '"'
inflate:         literal 'C'
inflate:         literal 'C'
inflate:         literal '0'
inflate:         literal ' '
inflate:         literal '1'
inflate:         literal '.'
inflate:         literal '0'
inflate:         literal '"'
inflate:         literal ';'
inflate:         literal ' '
inflate:         literal 'h'
inflate:         literal 't'
inflate:         literal 't'
inflate:         literal 'p'
inflate:         literal 's'
inflate:         literal ':'
inflate:         literal '/'
inflate:         literal '/'
inflate:         literal 'c'
inflate:         length 7
inflate:         distance 44
inflate:         literal 'c'
inflate:         length 6
inflate:         distance 43
inflate:         literal '.'
inflate:         literal 'o'
inflate:         literal 'r'
inflate:         literal 'g'
inflate:         literal '/'
inflate:         length 6
inflate:         distance 82
inflate:         length 6
inflate:         distance 81
inflate:         literal '/'
inflate:         literal 'z'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal 'o'
inflate:         literal '/'
inflate:         length 3
inflate:         distance 52
inflate:         literal '/'
inflate:         end of block
inflate:   check matches trailer
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 0x01
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x01
inflate:         length 258
inflate:         distance 1
inflate:         length 258
inflate:         distance 1
inflate:         length 258
inflate:         distance 1
inflate:         length 245
inflate:         distance 1
inflate:         literal 0x02
inflate:         literal 0x01
inflate:         literal 0x03
inflate:         literal 0x05
inflate:         literal 0x07
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 246
inflate:         distance 4
inflate:         literal 0x02
inflate:         literal 0x00
inflate:         literal 0x02
inflate:         literal 0x04
inflate:         literal 0x06
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 246
inflate:         distance 4
inflate:         length 258
inflate:         distance 1025
```

Here distance cannot be 1025 because `window (256) + output buffer (256) = 512`

Reply:
Actually, the `wbits` are different each time the stream is reset. I have made a commit which should fix it.

Reply:
So only way to fix the issue is to make sure inflate allocates larger buffer than twice the window size. Basically we would need 2048 bytes or more so it would still be 2^N.

Reply:
The problem is that `wsize` was never set to 0 after window freed (or before window needed to be reallocated). This caused it to have a non-zero value when it goes to `inflate_ensure_window` and thus never gets set to the proper value.

Reply:
Doesn't really matter why the window size was wrong, this is still a draft so it is supposed to have bugs... Only way to get the bugs fixed is to understand why something doesn't work correctly.